index.js

```diff
@@ -1037,15 +1037,18 @@ module.exports = function (log, indexesPath) {
   ) {
     seq = seq || 0
 
-    const sorted = sortedByTimestamp(bitset, descending)
+    let sorted = sortedByTimestamp(bitset, descending)
     const resultSize = sorted.size
 
-    const sliced = []
-    for (var i = 0; i < seq; ++i) sorted.poll()
-
-    while (!sorted.isEmpty()) {
-      sliced.push(sorted.poll())
-      if (limit !== null && sliced.length == limit) break
+    let sliced
+    if (seq === 0 && limit === 1) {
+      sliced = [sorted.peek()]
+    } else {
+      if (seq > 0) {
+        sorted = sorted.clone()
+        sorted.removeMany(() => true, seq)
+      }
+      sliced = sorted.kSmallest(limit || Infinity)
     }
```

Comment:
What happens in this block is that there is an optimization for the case of `seq=0`, `limit=1`, where we don't need to mutate the FPQ, and therefore we can eliminate the clone() and we don't need to do any add or remove (which are `O(log n)`) and instead we can use a `O(1)` operation, peek(). It doesn't happen that often, but might make startup a bit more pleasant.

The second block also avoids a `clone()` if `seq === 0`, because there's nothing to remove. So we postpone the clone() until it's absolutely necessary. I'm not sure how `removeMany` works, but here **I'm assuming that it runs through the elements in order**. I'm curious whether this is true and whether it works in the wild.

`kSmallest`, though, is a good find! It's actually not documented on the readme, but it's in the implementation. Comments in the code say:

> // return the k 'smallest' elements of the queue
// runs in O(k log k) time
// this is the equivalent of repeatedly calling poll, but
// it has a better computational complexity, which can be
// important for large data sets.

Reply:
That's a really nice find